### DISCUSSION TITLE
Multi-tenancy: account_login_filter

### DISCUSSION THREAD

**Discussion by pboling** [jeremyevans/rodauth#378] [General]
Did the idea for supporting multi-tenancy discussed in this [google groups thread](https://groups.google.com/g/rodauth/c/4bLB4HeiQ-E?pli=1) in 2021, which involved adding a new feature hypothetically called `account_login_filter`, ever become reality?  I am not able to find that exact method name in the code base, but perhaps when it was added it was named something else.

**Comment by janko** [comment]
The `account_table_ds` private method got added, which is shared by `account_from_login` and `account_from_session`, as well as retrieval of accounts from tokens. I believe you could use it for multi-tenancy. Taking the example from the google group thread:

```rb
auth_class_eval do
  private
  def account_table_ds
    super.where(tenant_id: request.env[:tenant_id])
  end
end
```

**Reply by pboling** [reply]
Within rodauth, how would one set `request.env[:tenant_id]`?  I'm looking at shared_vars plugin, and but I'm not sure if that it what I want.

**Reply by pboling** [reply]
`r.env[:tenant_id] = tenant_id` looks like the winner.

**Reply by pboling** [reply]
Followup question, also from the original Google Groups thread.  There was another example there, which was purported to be already supported.
```
    new_account do |_login|
      super.merge(tenant_id: request.env[:tenant_id])
    end
```
I am currently attempting to do exactly that, but it fails:
```
RuntimeError - implicit argument passing of super from method defined by define_method() is not supported. Specify all arguments explicitly.:
  app/misc/rodauth_main.rb:182:in `block (2 levels) in <class:RodauthMain>'
```
where line 182 is the `new_account` block with the super call just above.

**Reply by pboling** [reply]
Should I be using `before_create_account` instead to do this?  I have seen other examples in the Mastering Roda docs (IIRC) that made it look like that would also work.

**Reply by pboling** [reply]
OK, I have it all working (via `before_create_account`) up to the example code your gave, which is failing because `request` is not available in that context.
```
NoMethodError at /sd-api/auth/<my-tenant-id>/create-account
============================================================

undefined method `request' for nil:NilClass

> To access an interactive console with this error, point your browser to: /__better_errors


app/misc/rodauth_main.rb, line 194
----------------------------------

``` ruby
  191       auth_class_eval do
  192         private
  193         def account_table_ds
> 194           super.where(tenant_id: request.env[:tenant_id])
  195         end
  196       end
  197   
  198       # ==> Redirects
  199       # Redirect to home page after logout.
```

App backtrace
-------------

 - app/misc/rodauth_main.rb:194:in `account_table_ds'
 - app/mailers/rodauth_mailer.rb:58:in `block in rodauth'
 - app/mailers/rodauth_mailer.rb:58:in `rodauth'
 - app/mailers/rodauth_mailer.rb:5:in `verify_account'
 - app/misc/rodauth_main.rb:117:in `block (3 levels) in <class:RodauthMain>'
 - app/controllers/application_controller.rb:54:in `identity_cache_memoization'
 - app/controllers/application_controller.rb:49:in `set_time_zone'
 - app/misc/rodauth_app.rb:29:in `block (5 levels) in <class:RodauthApp>'
 - app/misc/rodauth_app.rb:21:in `block (4 levels) in <class:RodauthApp>'
 - app/misc/rodauth_app.rb:15:in `block (3 levels) in <class:RodauthApp>'
 - app/misc/rodauth_app.rb:14:in `block (2 levels) in <class:RodauthApp>'
 - app/misc/rodauth_app.rb:13:in `block in <class:RodauthApp>'
 - lib/vital_connect/middleware/omniauth_bypass.rb:50:in `call'

Full backtrace
--------------

 - rodauth (2.32.0) lib/rodauth/features/base.rb:149:in `request'
 - app/misc/rodauth_main.rb:194:in `account_table_ds'
 - rodauth (2.32.0) lib/rodauth/features/base.rb:772:in `account_ds'
 - app/mailers/rodauth_mailer.rb:58:in `block in rodauth'
 - app/mailers/rodauth_mailer.rb:58:in `rodauth'
 - app/mailers/rodauth_mailer.rb:5:in `verify_account'
 - actionpack (7.1.2) lib/abstract_controller/base.rb:224:in `process_action'
...
```

**Reply by pboling** [reply]
When I change the code to:
```
        super.where(tenant_id: scope.request.env[:tenant_id])
```
the error is the same, which is as expected, since `request` is effectively delegated to `scope`, and which means that `scope` is `nil`.

**Reply by jeremyevans** [reply]
This may be specific to rodauth-rails and not Rodauth itself, as I'm not aware of Rodauth itself creating a Rodauth instance with a nil scope.

**Reply by janko** [reply]
The mailer in rodauth-rails retrieves the account using a Rodauth object without a Roda scope, because it can be run in a background job, where there is no request context.

If the primary key (`id`) is unique across all tenants, then I would suggest just skipping the filter when there is no `request`:

```rb
def account_table_ds
  if request
    super.where(tenant_id: request.env[:tenant_id])
  else
    super
  end
end
```

Alternatively, you can pass the tenant ID down to each mailer method, and set the Roda scope with it:

```rb
class RodauthMain < Rodauth::Rails::Auth
  configure do
    create_reset_password_email do
      RodauthMailer.reset_password(account_id, request.env[:tenant_id], reset_password_key_value)
    end
    # ...
  end
end
```
```rb
class RodauthMailer < ApplicationMailer
  def reset_password(account_id, tenant_id, key)
    @rodauth = rodauth(account_id, tenant_id) { @reset_password_key_value = key }
    @account = @rodauth.rails_account

    mail subject: @rodauth.reset_password_email_subject
  end
  # ...
  private
  def rodauth(account_id, tenant_id)
    instance = RodauthApp.new({ tenant_id: tenant_id }).rodauth
    instance.instance_eval { @account = account_ds(account_id).first! }
    instance.instance_eval(&block) if block
    instance
  end
end
```

**Reply by pboling** [reply]
Thanks for that explanation @janko !  You've just made me realize why I need to do *something*, and that what I just pushed has a bug.  The URL that will be in the mailer for account activation won't have the "tenant_id" in it, and we are using that as a "path key" in the route/path for our multi-tenancy.  So I'll need to, at minimum, fix the `verify-account` path generation to include the tenant_id.  I'd have the same problem with any other emails that generate paths for links back to the API, and also for the UX path helpers.

Is there an idiomatic way to have path generation be account-context dependent, so the tenant_id can be derived and injected into the path?

**Reply by pboling** [reply]
It appears I will have to redefine the built-in paths, using the [`path` plugin](https://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/Path.html) and redefining them with my desired prefix, and dynamic tenant path key.

Here's what I have now for multi tenancy, which works perfectly when the right routes are called, with the exception of the incorrect paths used in the mailer templates.  Also I don't know how to programmatically generate any of these routes, as would be needed for the mailer templates, and for automated request specs.

```
class RodauthApp < Rodauth::Rails::App
  configure RodauthMain
  plugin :shared_vars
  plugin :symbol_matchers

  symbol_matcher :path_key, /([A-Z0-9_-]+)/xi

  route do |r|
    r.on("api") do
      r.on("auth") do
        r.on(:path_key) do |path_key|
          tenant = Tenant.fetch_by_path_key(path_key)

          r.on(proc { tenant }) do
            shared[:tenant] = tenant
            shared[:path_key] = path_key
            r.env[:pk_tenant_id] = shared[:tenant].id
            r.rodauth
          end
        end
      end
    end
  end
end
```

Is it insane to nest rodauth this way?

**Reply by janko** [reply]
Nesting looks fine, but the `shared_vars` Roda plugin probably isn't doing anything, because it's intended for sharing data between Roda apps, but you have only one in your Rails app. The path `plugin` would likely be an overkill as well, because it won't be used by Rodauth's route helpers anyway.
